src/comps/orth.cc

```diff
@@ -0,0 +1,115 @@
+/*
+TODO #1: only store the upper triangle of the gram matrix in gram_vec,
+so that it can be of size k*(k+1)/2 instead of k*k.
+*/
+
+#include <lapack.hh>
+#include <RandBLAS.hh>
+#include <RandLAPACK.hh>
+
+namespace RandLAPACK::comps::orth {
+
+// Perfoms a Cholesky QR factorization
+template <typename T> 
+int Orth<T>::CholQR(
+        int64_t m,
+        int64_t k,
+        std::vector<T>& Q // pointer to the beginning
+){
+        using namespace blas;
+        using namespace lapack;
+        
+        T* Q_gram_dat = RandLAPACK::comps::util::resize(k * (k - 1) / 2, this->Q_gram);
+        T* Q_dat = Q.data();
+
+        // Find normal equation Q'Q - Just the upper triangular portion        
+        sfrk(Op::NoTrans, Uplo::Upper, Op::Trans, k, m, 1.0, Q_dat, m, 0.0, Q_gram_dat);
+
+        // Positive definite cholesky factorization
+        if (pftrf(Op::NoTrans, Uplo::Upper, k, Q_gram_dat))
+                this->chol_fail = true; // scheme failure 
+
+        tfsm(Op::NoTrans, Side::Right, Uplo::Upper, Op::NoTrans, Diag::NonUnit, m, k, 1.0, Q_gram_dat, Q_dat, m);
+        
+       return 0;
+}
+
+template <typename T> 
+int Stab<T>::PLU(
+        int64_t m,
+        int64_t n,
+        std::vector<T>& A,
+        std::vector<int64_t>& ipiv
+){
+        using namespace lapack;
+
+        // Not using utility bc vector of int
+        if(ipiv.size() < n) 
+                ipiv.resize(n);
+
+        if(getrf(m, n, A.data(), m, ipiv.data()))
+                return 1; // failure condition
+        RandLAPACK::comps::util::swap_rows<T>(m, n, A, ipiv);
+        RandLAPACK::comps::util::get_L<T>(m, n, A);
+
+        return 0;
+}
+
+template <typename T> 
+int Orth<T>::HQR(
+        int64_t m,
+        int64_t n,
+        std::vector<T>& A,
+        std::vector<T>& tau
+){
+        // Done via regular LAPACK's QR
+        // tau The vector tau of length min(m,n). The scalar factors of the elementary reflectors (see Further Details).
+        // tau needs to be a vector of all 2's by default
+        using namespace lapack;
+
+        RandLAPACK::comps::util::resize(n, tau);
+
+        T* A_dat = A.data();
+	T* tau_dat = tau.data();
+        if(geqrf(m, n, A_dat, m, tau_dat))
+                return 1; // Failure condition
+        ungqr(m, n, n, A_dat, m, tau_dat);
+
+        return 0;
+}
+
+// GEQR lacks "unpacking" of Q
+template <typename T> 
+int Orth<T>::GEQR(
+        int64_t m,
+        int64_t n,
+        std::vector<T>& A,
+        std::vector<T>& tvec
+){
+        using namespace lapack;
+
+        tvec.resize(5);
+
+        T* A_dat = A.data();
+        T* tvec_dat = tvec.data();
+        
+        geqr(m, n, A_dat, m, tvec_dat, -1);
+        int64_t tsize = (int64_t) tvec[0]; 
+        tvec.resize(tsize);
+        geqr(m, n, A_dat, m, tvec.data(), tsize);
+
+        return 0;
```

Comment:
There should be a check for whether ``geqr`` succeeded or returned an error code.